File: common/src/main/java/cc/turtl/chiselmon/system/dashboard/DashboardServer.java
package cc.turtl.chiselmon.system.dashboard;

import cc.turtl.chiselmon.ChiselmonConstants;
import cc.turtl.chiselmon.system.dashboard.handler.StaticFileHandler;
import cc.turtl.chiselmon.system.dashboard.handler.api.*;
import cc.turtl.chiselmon.system.tracker.EncounterDatabase;
import com.sun.net.httpserver.HttpServer;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.util.concurrent.Executors;

/**
 * Super basic and lightweight HTTP server for the Chiselmon data dashboard.
 * <p>
 * Lifecycle:
 * <code>
 * <br>DashboardServer server = new DashboardServer(db, port);
 * <br>server.start();
 * <br>server.stop();
 * </code>
 */
public class DashboardServer {
    private static final String LOCALHOST = "127.0.0.1";

    private final EncounterDatabase db;
    private final int port;
    private HttpServer httpServer;
    private final long startTimeMs;

    public DashboardServer(EncounterDatabase db, int port) {
        this.db = db;
        this.port = port;
        this.startTimeMs = System.currentTimeMillis();
    }

    public void start() throws IOException {
        httpServer = HttpServer.create(new InetSocketAddress(LOCALHOST, port), 0);

        // Serves the HTML/CSS/JS for frontend
        httpServer.createContext("/", new StaticFileHandler());

        // Serves API endpoints which get data from the EncounterDatabase
        httpServer.createContext("/api/info", new InfoHandler(db));
        httpServer.createContext("/api/dimensions", new DimensionsHandler(db));
        httpServer.createContext("/api/stats", new StatsHandler(db));
        httpServer.createContext("/api/species", new SpeciesHandler(db));
        httpServer.createContext("/api/biomes", new BiomesHandler(db));
        httpServer.createContext("/api/encounters", new RecentEncountersHandler(db));
        httpServer.createContext("/api/timeline", new TimelineHandler(db));
        httpServer.createContext("/api/heatmap", new HeatmapHandler(db));

        // Executes on a single thread for now, should be fine for a simple local dashboard
        httpServer.setExecutor(Executors.newSingleThreadExecutor(r -> {
            Thread t = new Thread(r, "chiselmon-dashboard");
            t.setDaemon(true);
            return t;
        }));

        httpServer.start();
        ChiselmonConstants.LOGGER.info("Chiselmon Dashboard started at http://localhost:{}/", port);
    }

    public void stop() {
        if (httpServer != null) {
            httpServer.stop(0);
            ChiselmonConstants.LOGGER.info("Chiselmon Dashboard server stopped.");
        }
    }

    public long uptime() {
        if (httpServer != null) {
            return System.currentTimeMillis() - startTimeMs;
        }
        return 0;
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/client/system/dashboard/DashboardServer.kt
package cc.turtl.chiselmon.client.system.dashboard

import cc.turtl.chiselmon.client.system.dashboard.api.endpoints.BiomesHandler
import cc.turtl.chiselmon.client.system.dashboard.api.endpoints.ContextHandler
import cc.turtl.chiselmon.client.system.dashboard.api.endpoints.DimensionsHandler
import cc.turtl.chiselmon.client.system.dashboard.api.endpoints.HeatmapHandler
import cc.turtl.chiselmon.client.system.dashboard.api.endpoints.PlayerPosHandler
import cc.turtl.chiselmon.client.system.dashboard.api.endpoints.RecentEncountersHandler
import cc.turtl.chiselmon.client.system.dashboard.api.endpoints.SpeciesHandler
import cc.turtl.chiselmon.client.system.dashboard.api.endpoints.StatsHandler
import cc.turtl.chiselmon.client.system.dashboard.api.endpoints.TimelineHandler
import cc.turtl.chiselmon.core.ChiselmonConstants
import cc.turtl.chiselmon.client.system.tracker.EncounterDatabase
import com.sun.net.httpserver.HttpServer
import java.net.InetSocketAddress
import java.util.concurrent.Executors

/**
 * Super basic and lightweight HTTP server for the Chiselmon data dashboard.
 *
 * Lifecycle:
 * ```
 * val server = DashboardServer(db, port)
 * server.start()
 * server.stop()
 * ```
 */
class DashboardServer(private val db: EncounterDatabase, private val port: Int) {

    private var httpServer: HttpServer? = null
    private val startTimeMs: Long = System.currentTimeMillis()

    fun start() {
        val server = HttpServer.create(InetSocketAddress(LOCALHOST, port), 0).apply {
            // Serves the HTML/CSS/JS for frontend
            createContext("/", StaticFileHandler())

            // Serves API endpoints which get data from the EncounterDatabase
            createContext("/api/dimensions/", DimensionsHandler(db))
            createContext("/api/stats/", StatsHandler(db))
            createContext("/api/species/", SpeciesHandler(db))
            createContext("/api/biomes/", BiomesHandler(db))
            createContext("/api/encounters/", RecentEncountersHandler(db))
            createContext("/api/timeline/", TimelineHandler(db))
            createContext("/api/heatmap/", HeatmapHandler(db))
            createContext("/api/context/", ContextHandler(db))
            createContext("/api/playerpos/", PlayerPosHandler(db))

            // Executes on a single thread for now, should be fine for a simple local dashboard
            executor = Executors.newSingleThreadExecutor { r ->
                Thread(r, "chiselmon-dashboard").apply { isDaemon = true }
            }
        }

        server.start()
        httpServer = server
        ChiselmonConstants.LOGGER.info("Chiselmon Dashboard started at http://localhost:{}/", port)
    }

    fun stop() {
        httpServer?.let {
            it.stop(0)
            ChiselmonConstants.LOGGER.info("Chiselmon Dashboard server stopped.")
        }
    }

    fun uptime(): Long = if (httpServer != null) System.currentTimeMillis() - startTimeMs else 0

    companion object {
        private const val LOCALHOST = "127.0.0.1"
    }
}
